app/api/shorten.py

```diff
@@ -0,0 +1,257 @@
+
+    long_url = request.json.get('url')
+
+    if request.json.get('vanity_string') and not g.current_user.is_anonymous:
+        vanity_string = request.json.get('vanity_string')
+        short_url = UrlSaver.generate_and_save_urls(long_url, g.current_user,
+                                                    vanity_string)
+        return jsonify({'id': short_url.id,
+                        'short_url': site_url + short_url.short_url}), 201
+
+    user = g.current_user
+
+    short_url = UrlSaver.generate_and_save_urls(long_url, user)
+    return jsonify({'id': short_url.id,
+                    'short_url': site_url + short_url.short_url}), 201
+
+
+@api.route('/<shorturl>', strict_slashes=False)
+def get_url(shorturl):
+    """Redirect short_urls to their long_url version.
+
+    Given a short url, the code looks up the short url in database
+    and if found redirects to the long url path. if not found sends a 404
+    status_code.
+    """
+    short_url = ShortUrl.findurl(shorturl, request)
+    if short_url and short_url.deleted:
+        abort(404, "This URL has been deleted.")
+    elif short_url and not short_url.is_active:
+        abort(400, "URL is inactive.")
+    elif short_url and short_url.is_active:
+        return redirect(short_url.long_url.long_url, code=302)
+    else:
+        abort(404)
+
+
+@api.route('/short_url/<int:id>/change_longurl', methods=['PUT'],
+           strict_slashes=False)
+@auth.login_required
+def change_long_url(id):
+    """Update target url of a given short_url."""
+    check_authentication_with_token()
+
+    try:
+        valid_url(request.json)
+    except MultipleInvalid:
+        abort(400, "Incomplete or too many required keys provided.")
+    except BadRequest:
+        abort(400, "The request does not contain a body.")
+
+    try:
+        short_url = [short_url for short_url in g.current_user.short_urls
+                     if short_url.id == id and not short_url.deleted][0]
+    except IndexError:
+        abort(404, "You do not have any such URL.")
+
+    long_url = request.json.get('url')
+    result = short_url.change_long_url(long_url, g.current_user)
+    if isinstance(result, ShortUrl):
+        short_url = site_url + result.short_url
+        return jsonify({'message': "You have previously shortened this URL"
+                        " to %s." % short_url}), 400
+    return jsonify({'message': "Target url successfully changed to %s."
+                    % long_url}), 200
+
+
+@api.route('/user/short_urls', strict_slashes=False)
+@auth.login_required
+def get_user_short_urls():
+    """Return a list of a users service shortened URLs."""
+    check_authentication_with_token()
+
+    short_url = [short_url for short_url in g.current_user.short_urls]
+    if not short_url:
+        abort(404, "You are yet to shorten any URL.")
+    return jsonify({'short_url list':
+                    [{'Date_added': x.when, 'Times_visted': x.no_of_visits,
```

Comment:
You might want to have a specific format for your property names. either snake_case, camelCase, ProperCamelCase. But try to avoid using all